lib/3scale/backend/storage_async/client.rb

```diff
@@ -202,20 +73,61 @@ def initialize_client(opts)
         end
 
         def init_host_client(opts)
-          uri = URI(opts[:url] || '')
-          host = uri.host || DEFAULT_HOST
-          port = uri.port || DEFAULT_PORT
-
-          endpoint = Async::IO::Endpoint.tcp(host, port)
-          Async::Redis::Client.new(endpoint, limit: opts[:max_connections])
+          endpoint = make_redis_endpoint(opts)
+          protocol = make_redis_protocol(opts)
+          Async::Redis::Client.new(endpoint, protocol: protocol, limit: opts[:max_connections])
         end
 
         def init_sentinels_client(opts)
           uri = URI(opts[:url] || '')
           name = uri.host
           role = opts[:role] || :master
+          protocol = make_redis_protocol(opts)
+
+          Async::Redis::SentinelsClientACLTLS.new(name, opts[:sentinels], role, protocol, opts)
+        end
+
+        # Authenticated RESP2 if credentials are provided, RESP2 otherwise
+        def make_redis_protocol(opts)
```

Comment:
Is this code just new custom development, or copy from upstream?

Reply:
It's custom. The `async-redis` client is so simple we need to implement the most basic things like sending the credentials to the redis server when it requires authentication. We also have to handle the SSL connections ourselves and. All `private` methods in the `Client` class are there to deal with these things.

The code it's base in the examples from the repo:

https://github.com/socketry/async-redis/blob/main/examples/auth/protocol.rb
https://github.com/socketry/async-redis/tree/main/guides/getting-started

Reply:
Wow, that's sad that it doesn't even do the basic stuff...
Thanks for the info!